library/ssl_tls13_server.c

```diff
@@ -24,16 +24,776 @@
+     *     2  .  33   random bytes ( starting with 4 bytes of Unix time )
+     *    34  .  34   session id length ( 1 byte )
+     *    35  . 34+x  session id
+     *   35+x . 35+x  DTLS only: cookie length ( 1 byte )
+     *   36+x .  ..   DTLS only: cookie
+     *    ..  .  ..   ciphersuite list length ( 2 bytes )
+     *    ..  .  ..   ciphersuite list
+     *    ..  .  ..   compression alg. list length ( 1 byte )
+     *    ..  .  ..   compression alg. list
+     *    ..  .  ..   extensions length ( 2 bytes, optional )
+     *    ..  .  ..   extensions ( optional )
+     */
+
+    /* Needs to be updated due to mandatory extensions
+     * Minimal length ( with everything empty and extensions ommitted ) is
+     * 2 + 32 + 1 + 2 + 1 = 38 bytes. Check that first, so that we can
+     * read at least up to session id length without worrying.
+     */
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 38 );
+
+    /* ...
+     * ProtocolVersion legacy_version = 0x0303; // TLS 1.2
+     * ...
+     * with ProtocolVersion defined as:
+     * uint16 ProtocolVersion;
+     */
+    if( !( p[0] == MBEDTLS_SSL_MAJOR_VERSION_3 &&
+           p[1] == MBEDTLS_SSL_MINOR_VERSION_3 ) )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "Unsupported version of TLS." ) );
+        MBEDTLS_SSL_PEND_FATAL_ALERT( MBEDTLS_SSL_ALERT_MSG_PROTOCOL_VERSION,
+                                      MBEDTLS_ERR_SSL_BAD_PROTOCOL_VERSION );
+        ret = MBEDTLS_ERR_SSL_BAD_PROTOCOL_VERSION;
+        return ret;
+    }
+    p += 2;
+
+    /*
+     * Save client random
+     */
+    MBEDTLS_SSL_DEBUG_BUF( 3, "client hello, random bytes",
+                           p, MBEDTLS_SERVER_HELLO_RANDOM_LEN );
+
+    memcpy( &ssl->handshake->randbytes[0], p, MBEDTLS_SERVER_HELLO_RANDOM_LEN );
+    p += MBEDTLS_SERVER_HELLO_RANDOM_LEN;
+
+    /*
+     * Parse session ID
+     */
+    legacy_session_id_len = p[0];
+    p++;
+
+    if( legacy_session_id_len > 32 )
+    {
+        MBEDTLS_SSL_DEBUG_MSG( 1, ( "bad client hello message" ) );
+        return( MBEDTLS_ERR_SSL_DECODE_ERROR );
+    }
+
+    ssl->session_negotiate->id_len = legacy_session_id_len;
+
+    /* Note that this field is echoed even if
+     * the client's value corresponded to a cached pre-TLS 1.3 session
+     * which the server has chosen not to resume. A client which
+     * receives a legacy_session_id_echo field that does not match what
+     * it sent in the ClientHello MUST abort the handshake with an
+     * "illegal_parameter" alert.
+     */
+    MBEDTLS_SSL_DEBUG_BUF( 3, "client hello, session id",
+                           buf, legacy_session_id_len );
+
+    memcpy( &ssl->session_negotiate->id[0], p, legacy_session_id_len );
+    p += legacy_session_id_len;
+
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, 2 );
+    cipher_suites_len = MBEDTLS_GET_UINT16_BE( p, 0 );
+    p += 2;
+
+    MBEDTLS_SSL_CHK_BUF_READ_PTR( p, end, cipher_suites_len );
+
+    /* store pointer to ciphersuite list */
+    cipher_suites_start = p;
```

Comment:
cipher_suites_start is used to store the original start pointer of ciphersuite list. it will be used to search for a matching ciphersuite after extensions parsing.  cipher_suites  will be used to store the pre-configured the ciphersuit lists.

Reply:
Re-order the cipher-suite matching code, seems we still need store the pointer of p.

Reply:
not addressed

Reply:
As of 0a1b54e, `cipher_suites_start` is still used instead of `cipher_suites` the latter being aligned with our TLS 1.3 coding style.